File: __Viejo/Formularios/Agregar/CS/add_banco.vb
Public Class add_banco
    Private Sub add_banco_Load(sender As Object, e As EventArgs) Handles MyBase.Load
        'Cargo todos los paises
        cargar_combo(cmb_pais, "SELECT id_pais, pais FROM paises ORDER BY pais ASC", basedb, "pais", "id_pais")

        If edicion Or borrado Then
            txt_nombre.Text = edita_banco.nombre
            cmb_pais.SelectedValue = edita_banco.id_pais
            txt_bancoN.Text = edita_banco.n_banco
            chk_activo.Checked = edita_banco.activo

            chk_secuencia.Enabled = False
        End If

        If borrado Then
            txt_nombre.Enabled = False
            cmb_pais.Enabled = False
            txt_bancoN.Enabled = False
            chk_activo.Enabled = False

            chk_secuencia.Enabled = False
            cmd_ok.Enabled = False
            cmd_exit.Enabled = False
        End If

    End Sub

    Private Sub cmd_ok_Click(sender As Object, e As EventArgs) Handles cmd_ok.Click
        If txt_nombre.Text = "" Then
            MsgBox("El campo 'Nombre' es obligatorio y está vacio", vbExclamation + vbOKOnly, "Centrext")
            Exit Sub
        End If

        Dim b As New banco

        b.nombre = txt_nombre.Text
        b.id_pais = cmb_pais.SelectedValue
        b.n_banco = Val(txt_bancoN.Text)
        b.activo = chk_activo.Checked

        If edicion Then
            b.id_banco = edita_banco.id_banco
            If Not updatebanco(b) Then
                MsgBox("Hubo un problema al actualizar el banco, consulte con su programdor", vbExclamation + vbOKOnly, "Centrex")
                closeandupdate(Me)
            End If
        Else
            addbanco(b)
        End If

        If chk_secuencia.Checked Then
            txt_nombre.Text = ""
            cmb_pais.SelectedValue = 1
            txt_bancoN.Text = ""
            chk_activo.Checked = True
        Else
            closeandupdate(Me)
        End If
    End Sub

    Private Sub cmd_exit_Click(sender As Object, e As EventArgs) Handles cmd_exit.Click
        closeandupdate(Me)
    End Sub

    Private Sub cmb_pais_KeyPress(sender As Object, e As KeyPressEventArgs) Handles cmb_pais.KeyPress
        'e.KeyChar = ""
    End Sub
End Class

File: Formularios/Agregar/add_banco.cs
using System;
using System.Linq;
using System.Windows.Forms;

namespace Centrex
{
    public partial class add_banco : Form
    {
        public add_banco()
        {
            InitializeComponent();
        }

        private async void add_banco_Load(object sender, EventArgs e)
        {
            try
            {
                // ================================
                // 🔹 Cargar ComboBox de Países con EF
                // ================================
                using (var ctx = new CentrexDbContext())
                {
                    var queryPaises = ctx.PaisEntity
                        .OrderBy(p => p.Pais)
                        .Select(p => new
                        {
                            p.IdPais,
                            p.Pais
                        });

                    await CargarComboAsync(
                        cmb_pais,
                        queryPaises,
                        displayMember: "Pais",
                        valueMember: "IdPais",
                        soloActivos: true
                    );
                }

                // ================================
                // 🔹 Si estamos editando o borrando
                // ================================
                if (edicion || borrado)
                {
                    txt_nombre.Text = edita_banco?.Nombre ?? string.Empty;

                    if (edita_banco is not null)
                    {
                        cmb_pais.SelectedValue = edita_banco.IdPais;
                        txt_bancoN.Text = edita_banco.NBanco?.ToString() ?? string.Empty;
                        chk_activo.Checked = edita_banco.Activo;
                    }

                    chk_secuencia.Enabled = false;
                }

                // ================================
                // 🔹 Si estamos borrando, desactivar controles
                // ================================
                if (borrado)
                {
                    txt_nombre.Enabled = false;
                    cmb_pais.Enabled = false;
                    txt_bancoN.Enabled = false;
                    chk_activo.Enabled = false;
                    chk_secuencia.Enabled = false;
                    cmd_ok.Enabled = false;
                    cmd_exit.Enabled = false;
                }
            }
            catch (Exception ex)
            {
                Interaction.MsgBox($"Error al cargar formulario: {ex.Message}", MsgBoxStyle.Critical, "Centrex");
            }
        }

        private void cmd_ok_Click(object sender, EventArgs e)
        {
            try
            {
                if (string.IsNullOrEmpty(txt_nombre.Text))
                {
                    Interaction.MsgBox("El campo 'Nombre' es obligatorio.", MsgBoxStyle.Exclamation, "Centrex");
                    return;
                }

                var b = new BancoEntity
                {
                    Nombre = txt_nombre.Text.Trim(),
                    IdPais = Conversions.ToInteger(cmb_pais.SelectedValue),
                    NBanco = int.TryParse(txt_bancoN.Text, out var bancoNumero) ? bancoNumero : null,
                    Activo = chk_activo.Checked
                };

                // ================================
                // 🔹 Modo Edición
                // ================================
                if (edicion)
                {
                    b.IdBanco = edita_banco.IdBanco;
                    if (!bancos.UpdateBanco(b))
                    {
                        Interaction.MsgBox("Hubo un problema al actualizar el banco.", MsgBoxStyle.Exclamation, "Centrex");
                    }
                }
                else
                {
                    // ================================
                    // 🔹 Modo Alta
                    // ================================
                    bancos.AddBanco(b);
                }

                // ================================
                // 🔹 Reiniciar o cerrar
                // ================================
                if (chk_secuencia.Checked)
                {
                    txt_nombre.Text = "";
                    cmb_pais.SelectedIndex = -1;
                    txt_bancoN.Text = "";
                    chk_activo.Checked = true;
                }
                else
                {
                    closeandupdate(this);
                }
            }
            catch (Exception ex)
            {
                Interaction.MsgBox($"Error al guardar el banco: {ex.Message}", MsgBoxStyle.Critical, "Centrex");
            }
        }

        private void cmd_exit_Click(object sender, EventArgs e)
        {
            closeandupdate(this);
        }

        private void cmb_pais_KeyPress(object sender, KeyPressEventArgs e)
        {
            // Evita que el usuario escriba texto en el combo
            e.Handled = true;
        }
    }
}
